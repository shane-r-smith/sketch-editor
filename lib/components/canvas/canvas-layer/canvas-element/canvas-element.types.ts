import type { LineConfig } from "konva/lib/shapes/Line";
import type { SketchElement } from "../../../../domain/sketch-element";

export interface CanvasElementProps extends LineConfig {
  element: SketchElement;
}
