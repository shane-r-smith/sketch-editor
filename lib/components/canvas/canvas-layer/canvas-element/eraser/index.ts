export * from "./eraser";
export * from "./eraser.hooks";
