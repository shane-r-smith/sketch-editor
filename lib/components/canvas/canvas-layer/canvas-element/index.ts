export * from "./pen";
export * from "./eraser";

export * from "./canvas-element";
export * from "./canvas-element.types";
