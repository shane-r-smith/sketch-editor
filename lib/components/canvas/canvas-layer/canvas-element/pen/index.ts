export * from "./pen";
export * from "./pen.hooks";
