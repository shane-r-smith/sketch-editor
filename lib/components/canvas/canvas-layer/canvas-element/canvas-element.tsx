import { Eraser } from "./eraser";
import { Pen } from "./pen";
import type { CanvasElementProps } from "./canvas-element.types";

export function CanvasElement({ element, ...rest }: CanvasElementProps) {
  switch (element.tool) {
    case "PEN":
      return <Pen element={element} {...rest} />;
    case "ERASER":
      return <Eraser element={element} {...rest} />;
    default:
      return null;
  }
}
