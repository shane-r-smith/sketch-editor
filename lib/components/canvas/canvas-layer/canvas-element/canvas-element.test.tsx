import { render } from "@testing-library/react";
import { vi } from "vitest";
import { CanvasElement } from "./canvas-element";
import type { CanvasElementProps } from "./canvas-element.types";
import type { SketchElement } from "../../../../domain";

// Mock Pen and Eraser components
vi.mock("./pen", () => ({
  Pen: (props: CanvasElementProps) => (
    <div data-testid="pen">{JSON.stringify(props, undefined, 2)}</div>
  ),
}));
vi.mock("./eraser", () => ({
  Eraser: (props: CanvasElementProps) => (
    <div data-testid="eraser">{JSON.stringify(props, undefined, 2)}</div>
  ),
}));

describe("<CanvasElement />", () => {
  const baseProps: Omit<CanvasElementProps, "element"> = {};

  const baseElement: Omit<SketchElement, "key"> = {
    points: [0, 0, 10, 10],
    stroke: "#000000",
    strokeWidth: 2,
  };

  it("renders Pen when element.key is 'PEN'", () => {
    const element: SketchElement = {
      ...baseElement,
      key: "PEN",
      strokeColourType: "HEX",
    };
    const { container } = render(
      <CanvasElement element={element} {...baseProps} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders Eraser when element.key is 'ERASER'", () => {
    const element: SketchElement = {
      ...baseElement,
      key: "ERASER",
    };
    const { container } = render(
      <CanvasElement element={element} {...baseProps} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders null for unknown element.key", () => {
    const element = { key: "UNKNOWN" } as unknown as SketchElement;
    const { container } = render(
      <CanvasElement element={element} {...baseProps} />
    );
    expect(container).toMatchSnapshot();
  });
});
