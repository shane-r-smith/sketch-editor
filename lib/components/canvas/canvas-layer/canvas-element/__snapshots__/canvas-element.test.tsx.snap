// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CanvasElement /> > renders Eraser when element.key is 'ERASER' 1`] = `
<div>
  <div
    data-testid="eraser"
  >
    {
  "element": {
    "points": [
      0,
      0,
      10,
      10
    ],
    "stroke": "#000000",
    "strokeWidth": 2,
    "key": "ERASER"
  }
}
  </div>
</div>
`;

exports[`<CanvasElement /> > renders Pen when element.key is 'PEN' 1`] = `
<div>
  <div
    data-testid="pen"
  >
    {
  "element": {
    "points": [
      0,
      0,
      10,
      10
    ],
    "stroke": "#000000",
    "strokeWidth": 2,
    "key": "PEN",
    "strokeColourType": "HEX"
  }
}
  </div>
</div>
`;

exports[`<CanvasElement /> > renders null for unknown element.key 1`] = `<div />`;
